src/module/actor/sheet/spellcasting-dialog.ts

```diff
@@ -1,112 +1,115 @@
+class SpellcastingCreateAndEditDialog extends FormApplication<Embedded<SpellcastingEntryPF2e>> {
+    private actor: ActorPF2e;
 
-    constructor(options: Partial<ApplicationOptions>, private dialogOptions: SpellcastingDialogOptions) {
-        super(options);
+    constructor(object: ActorPF2e | Embedded<SpellcastingEntryPF2e>, options: Partial<FormApplicationOptions>) {
+        super(
+            object instanceof Actor ? createEmptySpellcastingEntry(object) : object.clone({}, { keepId: true }),
+            options
+        );
+        this.actor = object instanceof Actor ? object : object.actor;
     }
 
-    static override get defaultOptions(): ApplicationOptions {
+    static override get defaultOptions(): FormApplicationOptions {
         const options = super.defaultOptions;
         options.id = "spellcasting-dialog";
         options.classes = [];
         options.title = game.i18n.localize("PF2E.SpellcastingTypeLabel");
         options.template = "systems/pf2e/templates/actors/spellcasting-dialog.html";
         options.width = 300;
+        options.submitOnChange = true;
+        options.closeOnSubmit = false;
         return options;
     }
 
-    override activateListeners($html: JQuery): void {
-        const typeField = $html.find("[name=spellcasting-type]");
-        const traditionField = $html.find("[name=magic-tradition]");
-        const abilityField = $html.find("[name=ability]");
-        const flexibleField = $html.find("[name=flexible]");
-
-        const updateData = () => {
-            this.result.spellcastingType = String(typeField.val()) as PreparationType;
-            this.result.tradition = String(traditionField.val()) as MagicTradition | "";
-            this.result.ability = String(abilityField.val()) as AbilityString | "";
-            this.result.flexible = Boolean(flexibleField.prop("checked"));
-
-            if (this.result.spellcastingType === "ritual") {
-                this.result.tradition = "";
-                this.result.ability = "";
-            }
-            if (this.result.spellcastingType !== "prepared") {
-                this.result.flexible = false;
-            }
-        };
-
-        typeField.on("change", () => {
-            const wasRitual = this.result.spellcastingType === "ritual";
-            updateData();
-
-            if (wasRitual && this.result.spellcastingType !== "ritual") {
-                this.result.tradition = "arcane";
-                this.result.ability = "cha";
-            }
-
-            if (this.result.spellcastingType === "innate") {
-                this.result.ability = "cha";
-            }
-
-            this.render(true);
-        });
-
-        $html.find("[data-button=finish]").on("click", () => {
-            updateData();
-            this.dialogOptions.callback(this.result, $html);
-            this.close();
-        });
-    }
-
     override getData() {
```

Comment:
Can the return type be annotated here?

Reply:
It could, but given it is unlikely to be subclassed, its already any land. But I might as well.